### PR TITLE
ramips: add support for Hi-Link HLK-7688A IoT Gateway Mode

### PR DESCRIPTION
**Pull request by ewpa**
Please can support for Hi-Link HLK-7688A in IoT Gateway Mode be added.

### CODE REVIEW

File: target/linux/ramips/mt76x8/base-files/etc/board.d/01_leds [hunk 1]
```diff
@@ -42,6 +42,9 @@ hilink,hlk-7628n|\
 skylab,skw92a)
 	set_wifi_led "$boardname:green:wlan"
 	;;
+hilink,hlk-7688a-gateway)
+	set_wifi_led "hlk-7688a:green:wlan"
```
Comment (adschm) [inline target/linux/ramips/mt76x8/base-files/etc/board.d/01_leds:46]: Please use DT trigger instead.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x8000000>;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,57600";
+	};
+
+	buttonkeys {
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:22]: keys

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		led-boot = &wlan;
+		led-failsafe = &wlan;
+		led-running = &wlan;
+		led-upgrade = &wlan;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		wlan: status {
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:42]: So, which?

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+};
+
+&uart1 {
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:63]: flash@0

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:66]: Can this go faster?

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "ALL";
+				reg = <0x0 0x02000000>;
+				read-only;
+			};
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:77]: Please remove that, unless there is a specific reason for having it.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+			partition@1 {
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:79]: partition@0

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts (hunk 6 above)
Comment (ewpa) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:77]: Factory firmware had ALL partition.  Easily remove.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+/dts-v1/;
+
+#include "mt7628an.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "hilink,hlk-7688a-gateway", "hilink,hlk-7688a", "mediatek,mt7628an-soc";
+	model = "Hi-Link HLK-7688A IoT Gateway";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:11]: I'm not a big fan of creating one image per "mode". I'd prefer to create one image hilink,hlk-7688a or hilink,hlk-7688a-iot for both modes.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x8000000>;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,57600";
+	};
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:20]: This node can be dropped.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+	buttonkeys {
+		compatible = "gpio-keys";
+
+		wdt_rst {
+			label = "wdt_rst";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:26]: Reset?

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts (hunk 8 above)
Comment (ewpa) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:11]: Okay, let me think about this some more.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+#include "mt7628an.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "hilink,hlk-7688a-gateway", "hilink,hlk-7688a", "mediatek,mt7628an-soc";
+	model = "Hi-Link HLK-7688A IoT Gateway";
+
+	memory@0 {
```
Comment (981213) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:13]: memory node can be dropped. linux has the ability to detect it correctly.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts (hunk 5 above)
Comment (ewpa) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:66]: U-Boot runs the SPI clock at 32 MHz.  Practical limit seems to be around 66 MHz.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts (hunk 8 above)
Comment (ewpa) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:11]: Do you have any tips on the best method to disable the switch when the board is in "IoT Device Mode", i.e. with a single Ethernet (instead of the 1x WAN and 4x LAN of the "IoT Gateway Mode")?

**Comment by ewpa** [comment]
I have resolved all the above issues in the amended commit 3ae29b6.  Do let me know if that was not the correct method to update this pull request.  Note the HLK-7688A IoT Device mode does boot with this configuration, but is not fully tested.

File: target/linux/ramips/image/mt76x8.mk [hunk 12]
```diff
@@ -118,6 +118,15 @@ define Device/hilink_hlk-7628n
 endef
 TARGET_DEVICES += hilink_hlk-7628n
 
+define Device/hilink_hlk-7688a
+  BLOCKSIZE := 64k
```
Comment (adschm) [inline target/linux/ramips/image/mt76x8.mk:122]: Why is that needed?

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <66000000>;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:57]: Speed > 50 MHz typically requires `m25p,fast-read;`

**Comment by adschm** [comment]
> Signed-off-by: Ewan Parker <34030942+ewpa@users.noreply.github.com>

Is that address used by intention?

Apart from that, you didn't move the LED trigger to DTS so far.
I'm also missing ethernet/esw setup in DTS, including the MAC address.
Please also add an overview of MAC addresses to the commit message, more details here: https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts (hunk 8 above)
Comment (adschm) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts]: While I don't really understand what's that supposed to achieve (why would I have a separate mode which just provides less functionality, I think this is something to be achieved with config. OpenWrt does not aim at providing the exact mirror of the vendor concept _by default_, but just some reasonable configuration. So, you can just flash the official image twice and configure each of them based on the configuration you want to get.

Comment (ewpa) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts]: By disabling some of the Ethernet ports, the pins can be used for other purposes, such as I2C or SD Card.

File: target/linux/ramips/image/mt76x8.mk (hunk 12 above)
Comment (ewpa) [reply target/linux/ramips/image/mt76x8.mk:122]: I will remove.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts (hunk 13 above)
Comment (ewpa) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:57]: Thanks, I missed that.

**Comment by ewpa** [comment]
I plan to procure another board to check consistency of MAC address positions in the flash.

**Comment by ewpa** [comment]
I have amended the commit again and hope to have fixed everything that was outstanding.  In this commit I also enabled (and tested) I2C and SPI CS1.  When testing with another board I needed to lower the SPI speed to 40 MHz.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts [hunk 14]
```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:2]: drop dts-v1, this has been moved to mt7628an.dtsi

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+		led-boot = &led_wlan;
+		led-failsafe = &led_wlan;
+		led-upgrade = &led_wlan;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_wlan: wlan {
+			label = "hlk-7688a:green:wlan";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:33]: please drop the model: prefix here, just use "green:wlan"

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_pins>, <&spi_cs1_pins>;
+
+	flash@0 {
+		compatible = "winbond,w25q256", "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <40000000>;
+		m25p,fast-read;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:65]: I don't think m25p,fast-read is required below 50 MHz?

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+&ethernet {
+	mtd-mac-address = <&factory 0x28>;
+};
+
+&esw {
+	mediatek,portmap = <0x3e>;
+};
+
+&wmac {
+	mtd-mac-address = <&factory 0x4>;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:116]: If you use the default location for caldata, this address is read automatically.

File: target/linux/ramips/mt76x8/base-files/etc/board.d/02_network [hunk 18]
```diff
@@ -218,6 +219,7 @@ ramips_setup_macs()
 	zbtlink,zbt-we1226)
 		wan_mac=$(macaddr_add "$(mtd_get_mac_binary factory 0x2e)" 1)
 		;;
+	hilink,hlk-7688a|\
 	wiznet,wizfi630s)
 		label_mac=$(mtd_get_mac_binary factory 0x4)
 		wan_mac=$(mtd_get_mac_binary factory 0x28)
```
Comment (adschm) [inline target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:225]: Setting wan to 0x28 does not make any sense, as you already have set &ethernet to that address in DTS ...

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts (hunk 16 above)
Comment (ewpa) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:65]: Agreed, I can see no benefit in retaining that under 50 MHz.

File: target/linux/ramips/mt76x8/base-files/etc/board.d/02_network (hunk 18 above)
Comment (ewpa) [reply target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:225]: My mistake, WAN MAC is at offset 0x2e.  On my devices the factory MACs stored at both 0x28 and 0x2e happen to be the same value.

**Comment by ewpa** [comment]
I have pushed commit 2009e3162c to address all concerns.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+			gpios = <&gpio 38 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	aliases {
+		led-boot = &led_wlan;
+		led-failsafe = &led_wlan;
+		led-upgrade = &led_wlan;
+	};
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:26]: Please move aliases block before keys block.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+			label = "green:wlan";
+			gpios = <&gpio 44 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
+	};
+};
+
+&state_default {
+	spis {
+		groups = "spis";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:41]: node name should be gpio instead of spis, and groups should be wdt and wlan_an IMO?

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+};
+
+&spi0 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_pins>, <&spi_cs1_pins>;
+
+	flash@0 {
+		compatible = "winbond,w25q256", "jedec,spi-nor";
```
Comment (adschm) [inline target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:61]: Is the first entry required here?

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts (hunk 19 above)
Comment (ewpa) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:26]: Will do.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts (hunk 20 above)
Comment (ewpa) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:41]: I will change to `wdt` and `wled_an`.

File: target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts (hunk 21 above)
Comment (ewpa) [reply target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:61]: Not strictly required, happy to get rid of the winbond reference.

**Comment by ewpa** [comment]
I have pushed amended commit e7b237d6b2 to address the last three issues.

**Comment by alex9x** [comment]
Hi! Does someone know, why don`t work software reset on HLK-7688A?
4.xx have patch for 32mb flash with 0x66 0x99 sequence to back flash in 3byte address mode.
18.06.9 works well, but master or 19.x.x based on 5.x kernel fail to make soft reset.
Any help are welcome!

**Comment by ewpa** [comment]
> Hi! Does someone know, why don`t work software reset on HLK-7688A?
> 18.06.9 works well, but master or 19.x.x based on 5.x kernel fail to make soft reset.

I also experienced this on earlier commits on master.  Using this commit and kernel and reboot works (for me currently):

_OpenWrt SNAPSHOT, r14641-e7b237d6b2
Linux 5.4.69 #0 Sat Oct 31 09:29:54 2020 mips GNU/Linux_

**Comment by adschm** [comment]
Thanks, will be merged later today.
